
# Analyzing Used Car Listings on eBay Kleinanzeigen
We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

1. dateCrawled - When this ad was first crawled. All field-values are taken from this date.
2. name - Name of the car.
3. seller - Whether the seller is private or a dealer.
4. offerType - The type of listing
5. price - The price on the ad to sell the car.
6. abtest - Whether the listing is included in an A/B test.
7. vehicleType - The vehicle Type.
8. yearOfRegistration - The year in which which year the car was first registered.
9. gearbox - The transmission type.
10. powerPS - The power of the car in PS.
11. model - The car model name.
12. kilometer - How many kilometers the car has driven.
13. monthOfRegistration - The month in which which year the car was first registered.
14. fuelType - What type of fuel the car uses.
15. brand - The brand of the car.
16. notRepairedDamage - If the car has a damage which is not yet repaired.
17. dateCreated - The date on which the eBay listing was created.
18. nrOfPictures - The number of pictures in the ad.
19. postalCode - The postal code for the location of the vehicle.
20. lastSeenOnline - When the crawler saw this ad last online.


The aim of this project is to clean the data and analyze the included used car listings.

In [45]:
import pandas as pd
import numpy as np

In [46]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Our dataset contains 20 columns, most of which are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.

## Clean Columns

In [47]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make a few changes here:

1. Change the columns from camelcase to snakecase.
2. Change a few wordings to more accurately describe the columns.

In [48]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Initial Data Exploration and Cleaning
We'll start by exploring the data to find obvious areas where we can clean the data.

In [49]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Our initial observations:

1. There are a number of text columns where all (or nearly all) of the values are the same:
     a. seller
     b. offer_type
2. The num_photos column looks odd, we'll need to investigate this further.

In [50]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

It looks like the num_photos column has 0 for every column. We'll drop this column, plus the other two we noted as mostly one value.

In [51]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

There are two columns, price and auto, which are numeric values with extra characters being stored as text. We'll clean and convert these.

In [52]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [73]:
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Exploring Odometer and Price

In [75]:
autos["odometer_km"].value_counts()

150000    31028
125000     4960
100000     2110
90000      1696
80000      1396
70000      1199
60000      1137
50000      1008
5000        911
40000       801
30000       769
20000       763
10000       249
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [76]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)

(2333,)
count    4.802700e+04
mean     7.503289e+03
std      1.626210e+05
min      0.000000e+00
25%      1.150000e+03
50%      2.990000e+03
75%      7.400000e+03
max      2.732222e+07
Name: price, dtype: float64


0       1335
500      757
1500     696
2500     614
1200     606
1000     602
600      511
3500     480
800      470
2000     438
999      413
750      402
650      394
850      388
4500     375
900      374
300      373
700      373
2200     361
1100     355
Name: price, dtype: int64

Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.
<br>
There are 1,421 cars listed with $0 price - given that this is only 2% of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [77]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

27322222    1
12345678    1
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
Name: price, dtype: int64

In [78]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1335
1      150
2        2
3        1
5        2
8        1
9        1
10       6
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       6
35       1
Name: price, dtype: int64

There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

<br>
Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [79]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     46681.000000
mean       5977.716801
std        9177.909479
min           1.000000
25%        1250.000000
50%        3100.000000
75%        7500.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the date columns

There are a number of columns with date information:

1. date_crawled
2. registration_month
3. registration_year
4. ad_created
5. last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.
<br>
We'll explore each of these columns to learn more about the listings.

In [80]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [81]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025192
2016-03-06    0.014160
2016-03-07    0.036246
2016-03-08    0.033547
2016-03-09    0.033247
2016-03-10    0.032240
2016-03-11    0.032454
2016-03-12    0.036824
2016-03-13    0.015874
2016-03-14    0.036332
2016-03-15    0.034361
2016-03-16    0.029498
2016-03-17    0.031790
2016-03-18    0.012810
2016-03-19    0.034661
2016-03-20    0.038024
2016-03-21    0.037317
2016-03-22    0.032840
2016-03-23    0.032197
2016-03-24    0.029477
2016-03-25    0.031512
2016-03-26    0.032069
2016-03-27    0.030783
2016-03-28    0.034597
2016-03-29    0.034104
2016-03-30    0.033804
2016-03-31    0.031790
2016-04-01    0.033804
2016-04-02    0.035561
2016-04-03    0.038774
2016-04-04    0.036610
2016-04-05    0.013003
2016-04-06    0.003085
2016-04-07    0.001414
Name: date_crawled, dtype: float64

In [82]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-04-07    0.001414
2016-04-06    0.003085
2016-03-18    0.012810
2016-04-05    0.013003
2016-03-06    0.014160
2016-03-13    0.015874
2016-03-05    0.025192
2016-03-24    0.029477
2016-03-16    0.029498
2016-03-27    0.030783
2016-03-25    0.031512
2016-03-17    0.031790
2016-03-31    0.031790
2016-03-26    0.032069
2016-03-23    0.032197
2016-03-10    0.032240
2016-03-11    0.032454
2016-03-22    0.032840
2016-03-09    0.033247
2016-03-08    0.033547
2016-04-01    0.033804
2016-03-30    0.033804
2016-03-29    0.034104
2016-03-15    0.034361
2016-03-28    0.034597
2016-03-19    0.034661
2016-04-02    0.035561
2016-03-07    0.036246
2016-03-14    0.036332
2016-04-04    0.036610
2016-03-12    0.036824
2016-03-21    0.037317
2016-03-20    0.038024
2016-04-03    0.038774
Name: date_crawled, dtype: float64

In [83]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004113
2016-03-07    0.005377
2016-03-08    0.007476
2016-03-09    0.009768
2016-03-10    0.010690
2016-03-11    0.012382
2016-03-12    0.023757
2016-03-13    0.008654
2016-03-14    0.012660
2016-03-15    0.016002
2016-03-16    0.016281
2016-03-17    0.028084
2016-03-18    0.007219
2016-03-19    0.015617
2016-03-20    0.020629
2016-03-21    0.020587
2016-03-22    0.020844
2016-03-23    0.018359
2016-03-24    0.019687
2016-03-25    0.018937
2016-03-26    0.016795
2016-03-27    0.015638
2016-03-28    0.020694
2016-03-29    0.022086
2016-03-30    0.024614
2016-03-31    0.023628
2016-04-01    0.022943
2016-04-02    0.024657
2016-04-03    0.025149
2016-04-04    0.024121
2016-04-05    0.125404
2016-04-06    0.223324
2016-04-07    0.132752
Name: last_seen, dtype: float64


The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [84]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )


(74,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000043
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-27    0.000043
2016-02-01    0.000021
2016-02-02    0.000043
2016-02-05    0.000043
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000043
2016-02-14    0.000043
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000043
2016-02-19    0.000064
2016-02-20    0.000043
2016-02-21    0.000064
2016-02-22    0.000021
2016-02-23    0.000086
                ...   
2016-03-09    0.033290
2016-03-10    0.031962
2016-03-11    0.032754
2016-03-12    0.036653
2016-03-13    0.017180
2016-03-14    0.035004
2016-03-15    0.034168
2016-03-16    0.029927
2016-03-17    0.031469
2016-03-18    0.013496
2016-03-19    0.033611
2016-03-20    0.038067
2016-03-21 

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [85]:
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

## Dealing with Incorrect Registration Year Data
Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [86]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.0

In [87]:
# Many ways to select rows in a dataframe that fall within a value range for a column.
# Using `Series.between()` is one way.
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

## Exploring Price by Brand

In [88]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [89]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [90]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}


Of the top 5 brands, there is a distinct price gap:

1. Audi, BMW and Mercedes Benz are more expensive
2. Ford and Opel are less expensive
3. Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

## Exploring Mileage

In [91]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
audi                 9336
bmw                  8332
ford                 3749
mercedes_benz        8628
opel                 2975
volkswagen           5402

In [92]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [93]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [94]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

## Replacing German with English

In [95]:
autos["unrepaired_damage"].replace("nein", "No",inplace=True)
autos["unrepaired_damage"].replace("ja", "Yes",inplace=True)

In [96]:
autos["unrepaired_damage"].unique()

array(['No', nan, 'Yes'], dtype=object)

## How much cheaper are cars with damage than their non-damaged counterparts?

In [97]:
not_damaged= autos[autos["unrepaired_damage"]=="No"]
damaged = autos[autos["unrepaired_damage"]=="Yes"]

In [98]:
not_mean = not_damaged["price"].mean()
mean  = damaged["price"].mean()

In [99]:
print(not_mean,mean)

7164.033102796004 2241.146035242291


We can notice that damaged cars are very much cheaper than undamaged cars

## Find the most common brand/model combinations

In [100]:
autos.groupby(['brand','model']).size()

brand       model      
alfa_romeo  145               4
            147              80
            156              88
            159              32
            andere           60
            spider           32
audi        100              57
            200               1
            80              198
            90                8
            a1               82
            a2               42
            a3              825
            a4             1231
            a5              126
            a6              797
            a8               69
            andere          216
            q3               28
            q5               62
            q7               40
            tt              144
bmw         1er             521
            3er            2615
            5er            1132
            6er              30
            7er             126
            andere           38
            i3                1
            m_reihe          43
                

In [102]:
s = autos.groupby(['brand','model']).size()
df = s.loc[s.groupby(level=0).idxmax()].reset_index().drop(0,axis=1)
df

brand       model
0      alfa_romeo         156
1            audi          a4
2             bmw         3er
3       chevrolet      andere
4        chrysler      andere
5         citroen      andere
6           dacia     sandero
7          daewoo       matiz
8        daihatsu       cuore
9            fiat       punto
10           ford       focus
11          honda       civic
12        hyundai     i_reihe
13         jaguar      andere
14           jeep       grand
15            kia      andere
16           lada        niva
17         lancia     ypsilon
18     land_rover  freelander
19          mazda     3_reihe
20  mercedes_benz    c_klasse
21           mini      cooper
22     mitsubishi        colt
23         nissan       micra
24           opel       corsa
25        peugeot     2_reihe
26        porsche         911
27        renault      twingo
28          rover      andere
29           saab      andere
30           seat       ibiza
31          skoda     octavia
32          smart      fortwo
33         subaru      legacy
34         suzuki      andere
35         toyota       yaris
36        trabant         601
37     volkswagen        golf
38          volvo         v70